feat(organization): F4 - Internal dataset governance (PATHWAYS-809, HEXA-1349)

yolanfery opened the pull request:
Provide a way to share datasets across the whole organization

## Changes

- To avoid issue with conflicting dataset names, change the URL of datasets to include a `/from/`
- Adapt permissions
- added `sharedWithOrganization` to schema and model
- Filtering for users now include datasets form org sharing

## How/what to test

Datasets are correctly shared/viewable between workspaces

## Screenshots / screencast

<img width="2211" height="518" alt="Screenshot 2025-08-28 at 16 12 15" src="https://github.com/user-attachments/assets/e788049e-8d86-4d46-b7e4-4215efbcdd08" />
<img width="2211" height="518" alt="Screenshot 2025-08-28 at 16 12 21" src="https://github.com/user-attachments/assets/6b1b1d0e-ce7b-4c64-bd52-7b0ef020aaec" />

YolanFery commented on backend/hexa/datasets/schema/queries.py:
```diff
@@ -67,7 +67,7 @@ def resolve_dataset_link_by_slug(_, info, **kwargs):
         return DatasetLink.objects.filter_for_user(request.user).get(
             dataset__slug=kwargs["dataset_slug"],
-            workspace__slug=kwargs["workspace_slug"],
+            dataset__workspace__slug=kwargs["workspace_slug"],
```
Consider the source workspace

YolanFery commented on backend/hexa/user_management/permissions.py:
```diff
@@ -73,8 +73,8 @@ def archive_workspace(principal: User, organization: Organization):
 
-def list_all_workspaces(principal: User, organization: Organization):
-    """Only admin and owner users can list all workspaces"""
+def has_admin_privileges(principal: User, organization: Organization):
+    """Check if user has admin or owner privileges in the organization"""
```
Let's use that later to bypass restrictions (like access to workspace)

bramj commented on backend/hexa/datasets/models.py:
```diff
@@ -49,7 +55,14 @@ def filter_for_workspace_slugs(
+            | Q(  # 2. The dataset is shared with organization and user is member of the organization and a link exists to a workspace with slug in the list
+                shared_with_organization=True,
+                workspace__organization__organizationmembership__user=user,
+                links__workspace__slug__in=workspace_slugs,
```
I think we can extract some common behavior between `filter_for_user` and  `filter_for_workspace_slugs`, what do you think?

YolanFery replied:
DRYed it, thanks

bramj reviewed (changes requested):
There's some strange behavior on the following edge case:
1. Create a dataset in workspace A (Org A), share it with the organization A
2. In a different workspace B (Org A), share this same dataset with yet another workspace C

Now there are two possibilities:
1. Workspace C belongs to Org A: step 2 didn't make sense, I think we shouldn't allow that
2. Workspace C belongs to Org B: I suppose that makes sense, but it doesn't seem to work

All this is rather complex. Maybe we should disable the re-sharing of datasets that don't belong to one's organization? What do you think? // @mcrimi 

https://github.com/user-attachments/assets/a16dcc3c-784c-4a28-85e8-b9d9d6309773

One more thing: Creating a new dataset appears broken for me (when I click "Create" in the dialog, the UI doesn't react), but not sure if that's just me.

YolanFery commented on backend/hexa/datasets/permissions.py:
```diff
@@ -51,24 +58,37 @@ def download_dataset_version(principal: User, version: DatasetVersion):
         or dataset.workspace.workspacemembership_set.filter(
             user=principal,
             role__in=[WorkspaceMembershipRole.ADMIN, WorkspaceMembershipRole.EDITOR],
         ).exists()
```
Not sure at this point

YolanFery commented on backend/hexa/datasets/models.py:
```diff
@@ -40,7 +40,13 @@ def filter_for_user(self, user: AnonymousUser | User):
             return self._filter_for_user_and_query_object(
                 user,
-                models.Q(links__workspace__members=user),
+                Q(
```
Correct, not a priority but will be tackled as part of https://bluesquare.atlassian.net/browse/PATHWAYS-869

YolanFery commented on backend/hexa/datasets/models.py:
```diff
@@ -437,10 +451,29 @@ def filter_for_user(self, user: AnonymousUser | User):
+                        output_field=models.IntegerField(),
+                    )
+                )
+                .filter(is_original_link=1)
```
I simplified, it should be better now

YolanFery commented:
> There's some strange behavior on the following edge case:
> 
> 1. Create a dataset in workspace A (Org A), share it with the organization A
> 2. In a different workspace B (Org A), share this same dataset with yet another workspace C
> 
> Now there are two possibilities:
> 
> 1. Workspace C belongs to Org A: step 2 didn't make sense, I think we shouldn't allow that
> 2. Workspace C belongs to Org B: I suppose that makes sense, but it doesn't seem to work
> 

It was just a bug and I fixed it

> One more thing: Creating a new dataset appears broken for me (when I click "Create" in the dialog, the UI doesn't react), but not sure if that's just me.

Fixed it
